resources/aws/gateway.go

```diff
@@ -108,15 +110,45 @@ func (g *Gateway) Delete() error {
 		return microerror.MaskAny(err)
 	}
 
-	if _, err := g.Clients.EC2.DeleteInternetGateway(&ec2.DeleteInternetGatewayInput{
-		InternetGatewayId: gateway.InternetGatewayId,
-	}); err != nil {
+	detachOperation := func() error {
+		if _, err := g.Clients.EC2.DetachInternetGateway(&ec2.DetachInternetGatewayInput{
+			InternetGatewayId: gateway.InternetGatewayId,
+			VpcId:             aws.String(g.VpcID),
+		}); err != nil {
+			return microerror.MaskAny(err)
+		}
+		return nil
+	}
+	if err := backoff.Retry(detachOperation, backoff.NewExponentialBackOff()); err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	deleteOperation := func() error {
+		if _, err := g.Clients.EC2.DeleteInternetGateway(&ec2.DeleteInternetGatewayInput{
+			InternetGatewayId: gateway.InternetGatewayId,
+		}); err != nil {
+			return microerror.MaskAny(err)
+		}
+		return nil
+	}
+	if err := backoff.Retry(deleteOperation, backoff.NewExponentialBackOff()); err != nil {
 		return microerror.MaskAny(err)
 	}
 
 	return nil
 }
 
+func (g *Gateway) Get() error {
```

Comment:
Why is `Get` manipulating the internal state? When we discussed the interface I thought we agreed on lookup functions returning what they are supposed to get. This here looks like the lookup tries to find the actual gateway ID. So I would expect to see something like this without manipulating the internal ID. Having getters manipulating internal state is counterintuitive IMO. 

```
func (g *Gateway) GetID() (string, error)  
```